types/hapi__joi/index.d.ts

```diff
@@ -1918,7 +1918,8 @@ declare namespace Joi {
         /**
          * Generates a schema object that matches an object data type (as well as JSON strings that have been parsed into objects).
          */
-        object<TSchema = any>(schema?: SchemaMap<TSchema>): ObjectSchema<TSchema>;
+        // tslint:disable-next-line:no-unnecessary-generics
+        object<TSchema = any, T = TSchema>(schema?: SchemaMap<T>): ObjectSchema<TSchema>;
```

Comment:
I think it would be better to add type argument to `ObjectSchema.keys()` method, not to the object.

e.g.
```
        /**
         * Sets or extends the allowed object keys.
         */
        keys(schema?: SchemaMap<TSchema>): this;
        keys<TExtendSchema>(schema: SchemaMap<TExtend>): ObjectSchema<TSchema & TExtendSchema>;
```

Reply:
Good idea, but this will be a breaking change and I could not make it work. I defined the Overwrite type as our current version of TypeScript does not support it but Because of type inference user should provide the `TExtension` if they want type checking. for example here:
The `TSchema` is `any` so the returned type of first `keys` is `ObjectSchema<Overwrite<any, User>>` (which is `any`). So other keys will be `any`.
```ts
const userSchemaError = Joi.object().keys<User>({
    name: Joi.string().required(),
    family: Joi.string(),
    height: Joi.number(), // $ExpectError
}).keys<{ more: number }>({
    notMore: Joi.number,
});
```
As users usually start their types with a `Joi.object` I think current workaround is our best option. (Define all possible future fields using `TSchema` of `Joi.object<TSchema>`).

Reply:
Good Idea. I actually spent some time to implement this but due to typescript type inference user should provide the Schema generic parameter on every validator extension in order to force TS to use his/her types. IMO defining all possible future fields in the `Joi.object` is a safer pattern.